src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs

```diff
@@ -341,31 +341,48 @@ protected override void OnScriptFileChanged(FileSystemEventArgs e)
             // create a TraceWriter wrapper that can be exposed to Node.js
             var log = (ScriptFunc)(p =>
             {
-                var logObject = (IDictionary<string, object>)p;
-                string logType = (string)logObject["type"];
-                var logData = (IDictionary<string, object>)logObject["data"];
-
-                if (logType.Equals("trace", StringComparison.OrdinalIgnoreCase))
+                var logData = (IDictionary<string, object>)p;
+                string message = (string)logData["msg"];
+                if (message != null)
                 {
-                    TraceLevel level = (TraceLevel)logData["lvl"];
-                    var evt = new TraceEvent(level, (string)logData["msg"]);
+                    try
+                    {
+                        TraceLevel level = (TraceLevel)logData["lvl"];
+                        var evt = new TraceEvent(level, message);
 
-                    using (BeginFunctionScope(invocationContext))
+                        // Node captures the AsyncLocal value of the first invocation, which means that logs
+                        // are correlated incorrectly. Here we'll overwrite that value with the correct value
+                        // immediately before logging.
+                        using (BeginFunctionScope(invocationContext))
+                        {
+                            // TraceWriter already logs to ILogger
+                            traceWriter.Trace(evt);
+                        }
+                    }
+                    catch (ObjectDisposedException)
                     {
-                        // TraceWriter already logs to ILogger
-                        traceWriter.Trace(evt);
+                        // if a function attempts to write to a disposed
+                        // TraceWriter. Might happen if a function tries to
+                        // log after calling done()
                     }
                 }
-                else if (logType.Equals("metric", StringComparison.OrdinalIgnoreCase) &&
-                    logData.TryGetValue("name", out string name, ignoreCase: true) &&
-                    logData.TryGetValue("value", out object valueObject, ignoreCase: true))
+
+                return Task.FromResult<object>(null);
+            });
+
+            var metric = (ScriptFunc)(p =>
+            {
+                var metricObject = (IDictionary<string, object>)p;
+
```

Comment:
nit: since we have full control over `name`, `value`, and `properties` via `origMetric`, the `TryGetValue` calls are unnecessary in my opinion and add a bit of clutter